### DISCUSSION TITLE
Set Access property explicitly

### DISCUSSION THREAD

**Discussion by rvanbekkum** [StefanMaron/BusinessCentral.LinterCop#51] [New Rule]
It would be helpful if a code analysis rule could give a warning on setting the `Access` property explicitly.
This holds for all object types where this property can be added.
I think the default severity could be set to `Disabled`, as it is not something everyone might want to get warnings about, but it would be nice to make developers aware of the possibility of changing it from public to something else, have them think about the accessibility of the objects they create: "should these objects actually be part of the public API?"

**Comment by StefanMaron** [comment]
maybe a duplicate of #27 ?

**Reply by rvanbekkum** [reply]
No, although closely related, but with this I am requesting a code analysis rule that checks if the `Access` property is defined in the object.
So, what I would like is that all objects get an access modifier, e.g., `Access = Public;` even if that is the default, but then at least you know that that is intentional.

For issue #27, the AL XML Documentation extension covers that for the most part in VS Code, although it cannot be included in build pipelines as easily, as far as I know.

**Reply by StefanMaron** [reply]
Got it :)

**Comment by StefanMaron** [comment]
I will remove Page and XML Port as it is not supported for this property. Should I also add table fields?  
![image](https://user-images.githubusercontent.com/24838311/137187470-6def5af6-927e-4f00-9b5f-9a16591252c1.png)

https://docs.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/properties/devenv-access-property

**Reply by StefanMaron** [reply]
See latest commit, If ok, I will release the new version

**Reply by rvanbekkum** [reply]
Doesn't the property also exist for pages and XML-ports? I remember testing it on all object types with the latest AL compiler. I remember it wasn't valid for report objects. Can't check right away, as I'm not on my PC right now.

For table fields I'd personally rather see a separate rule.

**Reply by rvanbekkum** [reply]
Ok, I have checked. It's indeed also not applicable for pages and xmlports at the moment. 😊 Thanks for fixing that, good catch!

What do you think about applying it for table fields? I think it's typically something you want to set at the object level, and only set it explicitly to `Internal/Local` for table fields considering public is a good standard (haven't had a case where I'd want to change this to be non-public. Could be a consideration to have it as a setting in `LinterCop.json` or to have it in a separate rule so that you can enable or disable it independently, I think that is my preference.
What are your thoughts about that? 😃

**Reply by StefanMaron** [reply]
I also thought about field check disabled be default and can be activated via LinterCop.json not sure if this makes it more complicated as needed tho. On the other hand, splitting this into a separate rule that hardly will ever be activated might not make that much sense either

**Reply by StefanMaron** [reply]
as I write about it, I might prefer the setting then

**Reply by rvanbekkum** [reply]
Or ... have LC0011 only check it at object level for now. And if/when someone wants/needs it for field-level checks, then look into adding that as well? :)

**Reply by StefanMaron** [reply]
I see no disadvantage in adding it disabled by default, and I think it would be straight forward to align this rule with the documentation. I get that if field warning would be always enabled this adds a huge overhead compared to objects level only, hence the option to enable/disable field check

**Reply by rvanbekkum** [reply]
Me neither, I don't think it would be difficult to implement either. :) I agree that it would be better to have that disabled, as adding an explicit property value for all table fields would add a huge overhead like you say.

**Reply by StefanMaron** [reply]
New version is released!

**Reply by rvanbekkum** [reply]
Great! 😄
